### ISSUE TITLE
Ergonomics of logging & simulate-update loops

### ISSUE THREAD

**Issue by ariasanovsky** [ariasanovsky/azdopt#5]
- [x] make it easier to run $n$ simulations
  - ~~save models to file, load from file~~
    - too early for this
  - [x] process batches of final states instead of calling the model on individuals
- [x] make it easier to display transition data
  - prints for $s_0$, $(a_1, r_1, s_1), \dots, (a_t, r_t, s_t), g_\approx(s_t)$
  - [x] when are we we within a tolerance of our estimate for $\inf_s c(s)$?
  - ~~aggregate results into nice plots~~
    - too early as well
  - [x] real-time logging of loss functions, performance
  - ~~CLI for real-time adjustments~~
    - way too early for this
- [x] how to select roots for the next epoch, for each thread?
  - prioritize states by $c(s) - g_\theta(s)$
  - ~~prioritize state pairs by $\text{dist}(s, s')$~~
    - not needed -- each thread will select a root from its own search
    - initial roots should be from each other since they are sampled from the space
- [x] how to update a model with a batch of results?
  - consider the loss function $$\ell_\theta(s) = (g_{\text{exp}}(s) - g_\theta(s))^2 + c_h\cdot \mathbb{H}(p_{\text{exp}}(s, \cdot), p_\theta(s, \cdot)) + c_r\cdot \|\theta\|_2^2$$
    - here, $c_h, c_r$ are hyperparameters corresponding to our biases against cross entropy and model regularization, respectively
    - $\mathbb{H}(p, q) = -\left\langle p, \log q\right\rangle = -\sum_{x\in\mathcal{X}}p(x)\cdot\log(q(x))$ is the cross-entropy of discrete distributions $p, q$ over $\mathcal{X}$
    - $\|\cdot\|_2$ is the $L_2$-norm (applied above to the model's vector of weights and biases)

**Comment by ariasanovsky** [comment]
![color-log](https://github.com/ariasanovsky/azdopt/assets/14944402/296095d2-21a5-4fe4-a343-9fe144cf4134)

```rust
for (_, log) in trees {
    println!("{log}")
}
```
here `log: BasicGraphLog` displays:
- the tree's root graph string: time remaining & colored graph
- the cost of the root graph
- the (length 1) transition vector
  - a transition is an action (an edge to recolor) and the action's reward = old cost - new cost (positive = good)
- the final node type (`New(g)`: a new node will be added to the tree with value `g` (expected future gain from a prediction)

**Comment by ariasanovsky** [comment]
Mostly resolved with [recent commits](https://github.com/ariasanovsky/azdopt/commit/1e2fab1feb807fbf295c97747640e98dbb18a146)
